src/results/marginalization.rs

```diff
@@ -0,0 +1,75 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use hashbrown::HashMap;
+use pyo3::prelude::*;
+
+fn marginalize<T: std::ops::AddAssign + Copy>(
+    counts: HashMap<String, T>,
+    indices: Option<Vec<usize>>,
+) -> HashMap<String, T> {
+    let mut out_counts: HashMap<String, T> = HashMap::with_capacity(counts.len());
+    let clbit_size = counts
+        .keys()
+        .next()
+        .unwrap()
+        .replace(&['_', ' '][..], "")
+        .len();
+    let all_indices: Vec<usize> = (0..clbit_size).collect();
+    counts
+        .iter()
+        .map(|(k, v)| (k.replace(&['_', ' '][..], ""), *v))
+        .for_each(|(k, v)| match &indices {
+            Some(indices) => {
+                if all_indices == *indices {
+                    out_counts.insert(k, v);
+                } else {
+                    let key_arr = k.as_bytes();
+                    let new_key: String = indices
+                        .iter()
+                        .map(|bit| {
+                            let index = clbit_size - *bit - 1;
+                            match key_arr.get(index) {
+                                Some(bit) => *bit as char,
+                                None => '0',
+                            }
```

Comment:
Yeah, I'm completely fine with the output being nonsense, since the input is too.  It's good to see that we're not going to be emitting a panic from Rust though, which I think would be very confusing.

Reply:
Well fwiw if rust were to panic normally this leads to a `PanicException` from pyo3: https://pyo3.rs/main/doc/pyo3/panic/struct.PanicException.html so it doesn't hard crash the program.